src/server/generic_family.cc

```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
   (*cntx)->SendLong(match_cnt);
 }
 
+struct SortEntryEmpty {};
+
+struct SortEntryScore {
+  double score;
+};
+
+template <
+    // Whether alphanumeric sort will be used.
+    bool ALPHA,
+    // Deduced StorageKind type for optimal size.
+    container_utils::StorageKind STORAGE_KIND,
+    // Whether to store double score. It will be stored for non-alphanumeric sorts
+    // when our items are not trivailly convertiable to doubles (i.e. they are not integers)
+    bool STORE_SCORE = !ALPHA && STORAGE_KIND != container_utils::StorageKind::INTEGRAL>
+struct SortEntry : public std::conditional_t<STORE_SCORE, SortEntryScore, SortEntryEmpty> {
+  // Key type. If we sort alphanumerically, we store strings. Else we mimic
+  // the container storage type.
+  using KEY_TYPE =
+      std::conditional_t<ALPHA, std::string, container_utils::StorageType<STORAGE_KIND>>;
+  // Comparison type. std:string for alphanumeric sort, double otherwise.
+  using COMPARE_TYPE =
```

Comment:
when your runtime parameter alpha is true, then STORE_SCORE is false, your SortEntry has only a string.
this branch you can do at the beginning and I think if you do that, you kinda consolidate all the conditionals you put in all the levels. you still write the dedicated code to sort strings but you write it here as well, but now it's just divided between all these templates.

but if you just do this one `if` at the upper level for `alpha=true`, you probably end up with less lines of code and at least alpha=true branch will be much clearer. What do you think?

Reply:
It will in this case likely.

However where to divide? A full dedicated branch? So... sort & write is duplicated? Might be okay for now.

Later (in a second PR) SORT will be extended with external keys and there it'll be extremely inconvenient to have multiple branches because lots of hard logic will be duplicated. So there should be at least some `variant<vector<T?>>` that can be interfaced with from common code so that:
* T? can return its key
* T? will possibly store an external compare by value
* T? can be compared (without overhead for numeric sort)

We can drop support for integers at all because integer encoding is only used in small containers. It'll be beneficial only if you sort a container thousands of times... which is an unlikely use-case.

It could be even simpler, we can drop support for partial integer encoding (the `variant<string,int>` type). However, we should keep `Iterate` templated, because in case of a numerical sort we could convert a long into a string and then parse that string as a double 🤯 Parsing doubles is really slow... Converting integers to strings is probably not, because it fits into SSO.

This leaves us with:
- [string] for alphanumeric sort without external keys
- [string, double] for numeric sort with or without external keys
- [string, string] for alphanumeric sort with external key ** out of scope for this PR **

We could introduce just one variable - ALPHA - so we keep the conditional inheritance and Cmp functions (the `: conditional<ALPHA, Score, Empty>` part). Or even one common type [string, string, double] (key, cmp_key, cmp_double) that covers all cases for the long and short run ([string, double] in this PR). However we could no longer be on par with redis sort performance

Reply:
1. does `vector` represent a single item to sort in case of multiple GET statements (aka `sort s1 by w_* get w_* get f_* GET #`)? Or vector represents the list of items you sort?  

2. What's this "partial integer encoding" ? I am having a hard time correlating this name with SORT documentation.
3. I am inclined to drop the support for integers if it makes the code simpler.
4. Regarding SortEntry design. I do not understand why `key` is either string, int or a variant. which option in the spec the variant handles? What does it mean having a key and a score? to which option it correlates in the spec? 
5. 
`T? can return its key` - is it the original key (int, or string) covering SET, LIST, or ZSET field names?
`T? will possibly store an external compare by value` is it for `BY` statement, so BY fetches string or double (Controlled by ALPHA)? 
`T? can be compared (without overhead for numeric sort)`: you are referring to non-alpha case where your compare operator has a scalar (double) or a pair<by_score, string> or for alpha case where you compare a scalar (string) or a `pair<by_string, key_string>` ?